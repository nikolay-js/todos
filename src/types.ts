import {Dispatch} from "react";

// Store
export type Filter = string;

export interface ITask {
    id: number,
    text: string,
    isCompleted: boolean,
}

export interface State {
  taskText: string;
  tasks: ITask[];
  filter: Filter
}

// Actions
export enum ActionType {
  ADD_TASK = 'ADD_TASK',
  TEXT_TASK = 'TEXT_TASK',
  REMOVE_TASK = 'REMOVE_TASK',
  COMPLETE_TASK = 'COMPLETE_TASK',
  CHANGE_FILTER = 'CHANGE_FILTER'
}

interface IAddTaskAction {
    type: ActionType.ADD_TASK,
    payload: ITask,
}

interface ITextTaskAction {
    type: ActionType.TEXT_TASK,
    payload: string,
}

interface IRemoveTaskAction {
    type: ActionType.REMOVE_TASK,
}

interface ICompleteTaskAction {
    type: ActionType.COMPLETE_TASK,
    payload: {
        id: number,
    }
}

interface IChangeFilterAction {
    type: ActionType.CHANGE_FILTER,
    payload: {
        activeFilter: Filter,
    }
}

export type TaskActionTypes = IAddTaskAction | ITextTaskAction | IRemoveTaskAction | ICompleteTaskAction;
export type FilterActionType = IChangeFilterAction;

export type Action = TaskActionTypes | FilterActionType;

export type TContextState = {
  state: State;
  changeState: Dispatch<Action>
}
