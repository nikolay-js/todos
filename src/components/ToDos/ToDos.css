.todo-wrapper {
  max-width: 600px;
  margin: 0 auto;
  position: relative;
  background-color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2),
              0 25px 50px 0 rgba(0, 0, 0, 0.1);
}

.todo__title {
  font-size: 100px;
  text-align: center;
  font-weight: 300;
  color: rgba(175, 47, 47, 0.15);
}
