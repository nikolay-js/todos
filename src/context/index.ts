import { createContext } from 'react'
import { State, TContextState } from '../types'

const lsTasks = JSON.parse(localStorage.getItem('todo-list') || '{}');
export const initialState: State = {
    taskText: '',
    tasks: Object.keys(lsTasks).length > 0 ? lsTasks : [],
    filter: 'all'
  }

export const ContextApp = createContext<Partial<TContextState>>({});
