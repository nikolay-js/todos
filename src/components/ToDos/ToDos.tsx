import React, { useContext, useEffect } from 'react';
import ToDoInput from '../ToDoInput/ToDoInput';

import './ToDos.css'
import ToDoList from '../ToDoList/ToDoList';
import { ContextApp, initialState } from '../../context';
import { ITask } from '../../types';
import Footer from '../Footer/Footer';
import { add, changeText, remove, complete, change } from '../../reducers/tasksReducer';

const ToDos: React.FC = () => {
  const { state = initialState, changeState } = useContext(ContextApp);

  useEffect(() => {
    localStorage.setItem('todo-list', JSON.stringify(state.tasks));
  }, [state.tasks])

  const isTasksExist = state.tasks && state.tasks.length > 0;

  const handleInputChange = ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
    changeText(value)(changeState);
  }

  const addTask = ({ key }: React.KeyboardEvent<HTMLInputElement>) => {
    if (state.taskText && state.taskText.length > 3 && (!Boolean(key) || key === 'Enter')) {
      const task = {
        id: Date.now(),
        text: state.taskText,
        isCompleted: false,
      }
      add(task)(changeState);
      changeText('')(changeState);
    }
  }

  const removeTask = () => {
    remove()(changeState);
  }

  const completeTask = (id: number) => {
    complete(id)(changeState);
  }

  const changeFilters = (activeFilter: string) => {
    change(activeFilter)(changeState);
  }

  const filterTasks = (tasks: ITask[] = [], activeFilter: string = '') => {
    switch (activeFilter) {
      case 'completed':
        return tasks?.filter(task => task.isCompleted);
      case 'active':
        return tasks?.filter(task => !task.isCompleted);
      default:
        return tasks;
    }
  }

  const getActiveTasksCounter = (tasks: ITask[] = []) => tasks.filter(task => !task.isCompleted).length;
  const filteredTasks = filterTasks(state.tasks, state.filter);
  const taskCounter = getActiveTasksCounter(state.tasks);

  return (
    <div className="todo-wrapper">
      <h1 className="todo__title">todos</h1>
      <ToDoInput onKeyPress={addTask} onChange={handleInputChange} value={state.taskText} />
      {isTasksExist && <ToDoList completeTask={completeTask} tasksList={filteredTasks} />}
      {isTasksExist && <Footer changeFilter={changeFilters} amount={taskCounter} activeFilter={state.filter} removeTask={removeTask} />}
    </div>
  );
};

export default ToDos;
