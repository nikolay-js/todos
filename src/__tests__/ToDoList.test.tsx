import * as React from 'react';
import {ContextApp} from "../context";

import {shallow} from "enzyme";
import {cleanup, render} from "@testing-library/react";

import {State} from "../types";
import ToDos from '../components/ToDos/ToDos';

describe('<ToDoList />',() => {

    afterEach(cleanup);

    const testState: State = {
      taskText: '',
      tasks: [
        { id: 1, text: 'new task', isCompleted: true },
        { id: 2, text: 'new task2', isCompleted: false }],
      filter: ''
    }

    const Wrapper = () => {
        return (
            <ContextApp.Provider value={{state: testState}}>
                <ToDos/>
            </ContextApp.Provider>
            )
    }

    it('renders the component without changes', () => {
        const component = shallow(<Wrapper/>);
        expect(component).toMatchSnapshot();
    });

    it('should render right tasks length', async () => {

        const {container} = render(<Wrapper/>);
        expect(container.querySelectorAll('li')).toHaveLength(testState.tasks.length);
    });

})
