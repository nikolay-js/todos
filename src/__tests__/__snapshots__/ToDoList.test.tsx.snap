// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDoList /> renders the component without changes 1`] = `
<ContextProvider
  value={
    {
      "state": {
        "filter": "",
        "taskText": "",
        "tasks": [
          {
            "id": 1,
            "isCompleted": true,
            "text": "new task",
          },
          {
            "id": 2,
            "isCompleted": false,
            "text": "new task2",
          },
        ],
      },
    }
  }
>
  <ToDos />
</ContextProvider>
`;
